share/picongpu/unit/TWTSTight.cpp

```diff
@@ -0,0 +1,225 @@
+                    result(4u + valueIdx) = testBfield.calcTWTSFieldY(pos, time);
+                    result(5u + valueIdx) = testBfield.calcTWTSFieldZ(pos, time);
+                }
+            });
+    }
+};
+
+/** Test TWTSTight laser functions
+ *
+ * Compares the on host and on device computed result to analytical results.
+ *
+ */
+struct twtsTightNumberTest
+{
+    void operator()()
+    {
+        using namespace ::pmacc;
+        const templates::twtstight::EField testEfield = templates::twtstight::EField(
+            0.0,
+            800.0e-9,
+            30.0e-15,
+            2.5e-6,
+            5. * (PI / 180.),
+            1.0,
+            0.0,
+            false,
+            30. * (PI / 180.));
+        const templates::twtstight::BField testBfield = templates::twtstight::BField(
+            0.0,
+            800.0e-9,
+            30.0e-15,
+            2.5e-6,
+            5. * (PI / 180.),
+            1.0,
+            0.0,
+            false,
+            30. * (PI / 180.));
+        using float_T = templates::twtstight::float_T;
+
+        constexpr uint32_t numBlocks = 1;
+        constexpr uint32_t numThreadsPerBlock = 1;
+        constexpr uint32_t numThreads = numBlocks * numThreadsPerBlock;
+        constexpr uint32_t numValuesPerThread = 6;
+        constexpr uint32_t numValues = numThreads * numValuesPerThread;
+        const float3_64 pos = float3_64{1.0e-6, 1.0e-6, 100.0 * 1.5e-6};
+        const float_64 time = float_64(1.0e-15);
+
+        HostBuffer<float_T, 1u> resultHost(numValues);
+        DeviceBuffer<float_T, 1u> resultDevice(numValues);
+        resultDevice.setValue(float_T(0.0));
+
+        PMACC_LOCKSTEP_KERNEL(GenerateEvals<numThreadsPerBlock>{})
+            .template config<numThreadsPerBlock>(
+                numBlocks)(numValues, numValuesPerThread, pos, time, resultDevice.getDataBox());
+
+        resultHost.copyFrom(resultDevice);
+
+        auto res = resultHost.getDataBox();
+        const float_64 Ex = float_64(0.18329124052693974);
+        const float_64 Ey = float_64(-0.009402050968104002);
+        const float_64 Ez = float_64(0.1054028749666347);
+        const float_64 Bx = float_64(3.5299706879027803e-10);
+        const float_64 By = float_64(5.334111474127282e-11);
+        const float_64 Bz = float_64(-6.090365721598194e-10);
+        const float_T epsilon = float_T(5.0e-6);
+        const float_T epsilon2 = float_T(5.0e-15);
+        bool isCorrect = isApproxEqual(precisionCast<float_T>(Ex), res[0], epsilon)
+            && isApproxEqual(testEfield.calcTWTSFieldX(pos, time), res[0], epsilon2);
+        std::cerr << std::setprecision(std::numeric_limits<float_T>::digits10);
+        if(!isCorrect)
+            std::cerr << "Ex (analytical reference):   " << Ex << "  , EField.calcTWTSFieldX (Device):   " << res[0]
+                      << "  , EField.calcTWTSFieldX (Host):   " << testEfield.calcTWTSFieldX(pos, time) << std::endl;
+        REQUIRE(isCorrect);
```

Comment:
A few things about this:
- I would prefer to use `CHECK` over `REQUIRE`. The latter will stop your test case which means you're holding back on important information like "Oh, the x-direction failed but the y-direction passes. That's a hint!". (`REQUIRE` is more intended for pre-conditions the failure of which invalidates the whole test case.)
- I would prefer to have the two conditions per dimension and field in separate `CHECK` blocks immediately. If you do that, Catch2 is already smart enough to show you the values before the check and the equality that failed. It might be that Catch2 is not smart enough to see through the encapsulation inside a function. If so, I'd prefer to have that check inline.
- Afterwards, you can safely remove your debug statements because that information should be provided by Catch2 automatically as said above.
- After the above are taken care of, we can make E and B vectors as use a `for`-loop here. Something along the lines of
```C++
auto hostE = testEfield.getTWTSField_Normalized(...);
for (int i...) {
    CHECK(isApproxEqual(precisionCast<float_T>(E[i]), res[i], epsilon));
    CHECK(isApproxEqual(hostE[i], res[i], epsilon2));
    // ... same for B
}
```

Reply:
In case we add the add `a == Catch::Approx(b).margin(epsilon)`  we can remove the `std::err`, I am for it.

btw This line is triggering a compiler issue we should check if using `CHECK()` is solving the issue.